### ISSUE TITLE
Dockerfile confusion/suggestions

### ISSUE THREAD

**Issue by atruskie** [drakkan/sftpgo#190]
Hi!

First of all, I love the recent work. Auto initialization and an official docker image are great. Thanks.

I was using the previous example Dockerfiles (e.g. the [Alpine](https://github.com/drakkan/sftpgo/blob/master/docker/sftpgo/alpine/Dockerfile#L27) example), and while I recognize they were unsupported, the change in the official Dockerfile is jarring. 

Some points of confusion for me:

1. the link to the docs on the container is broken on the package page(https://github.com/users/drakkan/packages/container/package/sftpgo#installation)
![404](https://user-images.githubusercontent.com/1452968/95831546-84336a80-0d7c-11eb-95fc-46d6cb300ada.gif)

2. the package page makes it hard to find the docs. Each container can have a description set, but does not appear to have anything set. Perhaps we can change that?
![image](https://user-images.githubusercontent.com/1452968/95831682-b04eeb80-0d7c-11eb-9268-8fb68cc85f8c.png)

3. The previous docker image contained what I thought was a fairly simple and self-explanatory definition of mount points
  
  https://github.com/drakkan/sftpgo/blob/5b79379c9038359d2aecb7dda6f41db5533901e4/docker/sftpgo/alpine/Dockerfile#L27
    
Whereas the new version I find quite confusing. Here is a translation table:

| Setting                              | Default             | Unsupported Dockerfile | New Dockerfile             |
|--------------------------------------|---------------------|------------------------|----------------------------|
| SFTPGO_CONFIG_DIR                    | .                   | /srv/sftpgo/config     | /etc/sftpgo                |
| SFTPGO_DATA_PROVIDER__USERS_BASE_DIR | $HOME               | /data                  | /var/lib/sftpgo/users      |
| SFTPGO_HTTPD__BACKUPS_PATH           | $CONFIG_DIR/backups | /srv/sftpgo/backups    | /var/lib/sftpgo/backups    |
| SFTPGO_SFTPD__HOST_KEYS              | $CONFIG_DIR         | /srv/sftpgo/config/    | /var/lib/sftpgo/host_keys/ |
| SFTPGO_DATA_PROVIDER__CREDENTIALS_PATH  | $CONFIG_DIR/credentials         | /srv/sftpgo/config/credentials    | /var/lib/sftpgo/credentials/ |

- There is only [one volume mount](https://github.com/drakkan/sftpgo/blob/master/Dockerfile.alpine#L73) for the new Dockerfile 
   - this makes it hard to bind config and data folders separately
   - I definitely want the user base dir mounted elsewhere... it points to very large network file system on the host
   - And I want to manage the rest of `/var/lib/sftpgo` with a docker volume
- It doesn't make sense to have _user data_ stored in `/var/lib` anyway. That directory is for [application specific state](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html)

4. The [assumption that I don't want to insert a custom config](https://github.com/drakkan/sftpgo/tree/master/docker#configuration) file and use only environment variables is misguided
5. There are many defaults set in the dockerfile that lead to confusion for me, especially since they don't match the defaults for sftpgo (the disparity between /etc/config and /var/lib)
6. Environment vars defaults make configuration with a config file difficult, since environment vars take precedence

---

Constructive suggestion (not tested):
https://github.com/drakkan/sftpgo/commit/743f45ad29e28ff1d9597ffd543d35b1673c70d4#diff-f865864730d7062c04cc13b1b85ba03ed5dbef3fa02fcb50b87d9117d0af1f4e

https://github.com/drakkan/sftpgo/blob/743f45ad29e28ff1d9597ffd543d35b1673c70d4/Dockerfile.alpine#L47-L70

Example usage:

```
docker run --name some-sftpgo \
    -p 127.0.0.1:8080:8090 \
    -p 2022:2022 \
    --mount type=bind,source=/my/own/sftpgodata,target=/data \
    --mount type=volume,target=/etc/sftpgo \
    -e SFTPGO_HTTPD__BIND_PORT=8090 \
    -d "ghcr.io/drakkan/sftpgo:edge"
```
Or:

```
docker run --name some-sftpgo \
    -p 127.0.0.1:8080:8080 \
    -p 2022:2022 \
    --mount type=bind,source=/my/own/sftpgodata,target=/data \
    --mount type=bind,source=/my_config_dir,target=/etc/sftpgo \
    -d "ghcr.io/drakkan/sftpgo:edge"
```

To summarize, simply do less.
- It creates less unexpected behavior compared with running sftpgo outside a container.
- It means users with other setups don't have to undo work in the dockerfile

**Comment by drakkan** [comment]
Hi,

thank you for your detailed explanation, I have to say that you are right for most things, I have probably oversimplified the things,

please see below some other answers

> Hi!
> 
> First of all, I love the recent work. Auto initialization and an official docker image are great. Thanks.
> 
> I was using the previous example Dockerfiles (e.g. the [Alpine](https://github.com/drakkan/sftpgo/blob/master/docker/sftpgo/alpine/Dockerfile#L27) example), and while I recognize they were unsupported, the change in the official Dockerfile is jarring.
> 
> Some points of confusion for me:
> 
> 1. the link to the docs on the container is broken on the package page(https://github.com/users/drakkan/packages/container/package/sftpgo#installation)
>    ![404](https://user-images.githubusercontent.com/1452968/95831546-84336a80-0d7c-11eb-95fc-46d6cb300ada.gif)

I linked the main README (as suggested by GitHub) but all relative links are now broken

> 2. the package page makes it hard to find the docs. Each container can have a description set, but does not appear to have anything set. Perhaps we can change that?
>    ![image](https://user-images.githubusercontent.com/1452968/95831682-b04eeb80-0d7c-11eb-9268-8fb68cc85f8c.png)

No sorry that description is auto-filled and I reported the issue [here](https://github.community/t/package-details-and-manifest-are-missing/136818). Maybe the best thing to do is to move the images to Docker Hub.

> 3. The previous docker image contained what I thought was a fairly simple and self-explanatory definition of mount points
> 
> https://github.com/drakkan/sftpgo/blob/5b79379c9038359d2aecb7dda6f41db5533901e4/docker/sftpgo/alpine/Dockerfile#L27
> 
> Whereas the new version I find quite confusing. Here is a translation table:
> 
> Setting	Default	Unsupported Dockerfile	New Dockerfile
> SFTPGO_CONFIG_DIR	.	/srv/sftpgo/config	/etc/sftpgo
> SFTPGO_DATA_PROVIDER__USERS_BASE_DIR	$HOME	/data	/var/lib/sftpgo/users
> SFTPGO_HTTPD__BACKUPS_PATH	$CONFIG_DIR/backups	/srv/sftpgo/backups	/var/lib/sftpgo/backups
> SFTPGO_SFTPD__HOST_KEYS	$CONFIG_DIR	/srv/sftpgo/config/	/var/lib/sftpgo/host_keys/
> SFTPGO_DATA_PROVIDER__CREDENTIALS_PATH	$CONFIG_DIR/credentials	/srv/sftpgo/config/credentials	/var/lib/sftpgo/credentials/
> * There is only [one volume mount](https://github.com/drakkan/sftpgo/blob/master/Dockerfile.alpine#L73) for the new Dockerfile
>   
>   * this makes it hard to bind config and data folders separately
>   * I definitely want the user base dir mounted elsewhere... it points to very large network file system on the host
>   * And I want to manage the rest of `/var/lib/sftpgo` with a docker volume
> * It doesn't make sense to have _user data_ stored in `/var/lib` anyway. That directory is for [application specific state](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html)
> 

yes you are right, we should use `/srv/sftpgo/data`

> 1. The [assumption that I don't want to insert a custom config](https://github.com/drakkan/sftpgo/tree/master/docker#configuration) file and use only environment variables is misguided

Most docker users want to use env var, anyway we should support config file too, so instead of defining env vars inside the Dockerfile we should change the config file using sed

> 2. There are many defaults set in the dockerfile that lead to confusion for me, especially since they don't match the defaults for sftpgo (the disparity between /etc/config and /var/lib)
> 3. Environment vars defaults make configuration with a config file difficult, since environment vars take precedence
> 
> Constructive suggestion (not tested):
> [743f45a#diff-f865864730d7062c04cc13b1b85ba03ed5dbef3fa02fcb50b87d9117d0af1f4e](https://github.com/drakkan/sftpgo/commit/743f45ad29e28ff1d9597ffd543d35b1673c70d4#diff-f865864730d7062c04cc13b1b85ba03ed5dbef3fa02fcb50b87d9117d0af1f4e)
> 
> https://github.com/drakkan/sftpgo/blob/743f45ad29e28ff1d9597ffd543d35b1673c70d4/Dockerfile.alpine#L47-L70
> 
> Example usage:
> 
> ```
> docker run --name some-sftpgo \
>     -p 127.0.0.1:8080:8090 \
>     -p 2022:2022 \
>     --mount type=bind,source=/my/own/sftpgodata,target=/data \
>     --mount type=volume,target=/etc/sftpgo \
>     -e SFTPGO_HTTPD__BIND_PORT=8090 \
>     -d "ghcr.io/drakkan/sftpgo:edge"
> ```
> 
> Or:
> 
> ```
> docker run --name some-sftpgo \
>     -p 127.0.0.1:8080:8080 \
>     -p 2022:2022 \
>     --mount type=bind,source=/my/own/sftpgodata,target=/data \
>     --mount type=bind,source=/my_config_dir,target=/etc/sftpgo \
>     -d "ghcr.io/drakkan/sftpgo:edge"
> ```
> 
> To summarize, simply do less.
> 
> * It creates less unexpected behavior compared with running sftpgo outside a container.
> * It means users with other setups don't have to undo work in the dockerfile

I'll try to send a pull request later today or during the next week-end, thank you

**Comment by atruskie** [comment]
Thanks for responding.

> I linked the main README (as suggested by GitHub) but all relative links are now broken

Bummer - I figured it was something like that

> Most docker users want to use env var, anyway we should support config file too, so instead of defining env vars inside the Dockerfile we should change the config file using sed

I might disagree with that assertion. Docker Swarm supports config files through [Docker Configs](https://docs.docker.com/engine/swarm/configs/#about-configs) and Kuberenetes has [ConfigMaps](https://kubernetes.io/docs/concepts/configuration/configmap/) which can mount configs as volumes or environment variables.

Don't get me wrong, environment variables are popular too. You just don't need to set them as defaults.

---

I'm happy to help/contribute. Just let me know if you want me to do anything.

**Comment by drakkan** [comment]
Hi,

I propose the following changes:

```
diff --git a/Dockerfile.alpine b/Dockerfile.alpine
index 6e79ebd..d02faab 100644
--- a/Dockerfile.alpine
+++ b/Dockerfile.alpine
@@ -36,7 +36,7 @@ SHELL ["/bin/bash", "-c"]
 # https://github.com/gliderlabs/docker-alpine/issues/367#issuecomment-424546457
 RUN test ! -e /etc/nsswitch.conf && echo 'hosts: files dns' > /etc/nsswitch.conf
 
-RUN mkdir -p /etc/sftpgo /var/lib/sftpgo /usr/share/sftpgo
+RUN mkdir -p /etc/sftpgo /var/lib/sftpgo /usr/share/sftpgo /srv/sftpgo
 
 RUN addgroup -g 1000 -S sftpgo && \
     adduser -u 1000 -h /var/lib/sftpgo -s /sbin/nologin -G sftpgo -S -D -H sftpgo
@@ -44,32 +44,30 @@ RUN addgroup -g 1000 -S sftpgo && \
 # Install some optional packages used by SFTPGo features
 RUN apk add --update --no-cache rsync git
 
-# Override some configuration details
-ENV SFTPGO_CONFIG_DIR=/etc/sftpgo
-ENV SFTPGO_LOG_FILE_PATH=""
-ENV SFTPGO_HTTPD__TEMPLATES_PATH=/usr/share/sftpgo/templates
-ENV SFTPGO_HTTPD__STATIC_FILES_PATH=/usr/share/sftpgo/static
-
-# Sane defaults, but users should still be able to override this from env vars
-ENV SFTPGO_DATA_PROVIDER__USERS_BASE_DIR=/var/lib/sftpgo/users
-ENV SFTPGO_DATA_PROVIDER__CREDENTIALS_PATH=/var/lib/sftpgo/credentials
-ENV SFTPGO_HTTPD__BACKUPS_PATH=/var/lib/sftpgo/backups
-ENV SFTPGO_SFTPD__HOST_KEYS=/var/lib/sftpgo/host_keys/id_rsa,/var/lib/sftpgo/host_keys/id_ecdsa
-ENV SFTPGO_HTTPD__BIND_ADDRESS=""
-
 COPY --from=builder /workspace/sftpgo.json /etc/sftpgo/sftpgo.json
 COPY --from=builder /workspace/templates /usr/share/sftpgo/templates
 COPY --from=builder /workspace/static /usr/share/sftpgo/static
 COPY --from=builder /workspace/sftpgo /usr/local/bin/
 
-RUN sed -i "s|sftpgo.db|/var/lib/sftpgo/sftpgo.db|" /etc/sftpgo/sftpgo.json
+# Override some CLI flags via env vars
+ENV SFTPGO_LOG_FILE_PATH=""
+
+# Modify the default configuration file
+RUN sed -i "s|sftpgo.db|/var/lib/sftpgo/sftpgo.db|" /etc/sftpgo/sftpgo.json && \
+    sed -i "s|\"users_base_dir\": \"\",|\"users_base_dir\": \"/srv/sftpgo/data\",|" /etc/sftpgo/sftpgo.json && \
+    sed -i "s|\"templates\"|\"/usr/share/sftpgo/templates\"|" /etc/sftpgo/sftpgo.json && \
+    sed -i "s|\"static\"|\"/usr/share/sftpgo/static\"|" /etc/sftpgo/sftpgo.json && \
+    sed -i "s|\"backups\"|\"/srv/sftpgo/backups\"|" /etc/sftpgo/sftpgo.json && \
+    sed -i "s|\"credentials\"|\"/var/lib/sftpgo/credentials\"|" /etc/sftpgo/sftpgo.json && \
+    sed -i "s|\"bind_address\": \"127.0.0.1\",|\"bind_address\": \"\",|" /etc/sftpgo/sftpgo.json
 
-RUN chown -R sftpgo:sftpgo /etc/sftpgo && chown sftpgo:sftpgo /var/lib/sftpgo && \
+RUN chown -R sftpgo:sftpgo /etc/sftpgo && chown sftpgo:sftpgo /var/lib/sftpgo /srv/sftpgo && \
     chmod 640 /etc/sftpgo/sftpgo.json && \
-    chmod 750 /etc/sftpgo /var/lib/sftpgo
+    chmod 750 /etc/sftpgo /var/lib/sftpgo /srv/sftpgo
 
+WORKDIR /var/lib/sftpgo
 USER sftpgo
 
-VOLUME /var/lib/sftpgo
+VOLUME [ "/var/lib/sftpgo", "/srv/sftpgo" ]
 
 CMD sftpgo serve
```

This way:

1) we ship an already modified config file in `/etc/sftpgo.json`, no more env vars inside Dockerfile
2) we have 2 volume: `/var/lib/sftpgo` and `/srv/sftpgo`
3) sftpgo has `/var/lib/sftpgo` as home dir and it stores its files in /srv/sftpgo
4) the default home dir for SFTP users is `/srv/sftpgo/data`
5) backups path is `/srv/sftpgo/backups`
6) the daemon use `/var/lib/sftpgo` as home dir and as working dir

you can override default config file putting your `sftpgo.json` in `/var/lib/sftpgo/sftpgo.json` or better `/var/lib/sftpgo/.config/sftpgo/sftpgo.json` or using env vars only. Host keys are autogenerated inside `/var/lib/sftpgo` if not found.

This layout is quite common, for example the minio package on Arch Linux has:

- `/var/lib/minio` as home dir and certs files are there
- `/srv/minio` to store data

Will these changes fix your concerns?

I will also push images to DockerHub and after all these changes are merged I'll release 1.1.1.

I would be very grateful if you could help me to update the documentation. Thank you!

**Comment by drakkan** [comment]
we only should set these env vars:

```
ENV SFTPGO_HTTPD__TEMPLATES_PATH=/usr/share/sftpgo/templates
ENV SFTPGO_HTTPD__STATIC_FILES_PATH=/usr/share/sftpgo/static
```

since `templates` and `static` files are inside the image and changing them will prevent the image to start

**Comment by atruskie** [comment]
Looks great 😄. I'm a little concerned that the default config file is so different from the standard sftpgo config. Perhaps (if you think it worthwhile) it might be good to change the default directories in sftpgo to be relative to some app data path by default.

E.g. `credentials`, `host_keys` (and local database paths) are relative to new `app_data_dir` config setting, which defaults to be equal to `config_dir` (for backwards compatibility). Then, we'd need to only change the new `app_data_dir` setting (which removes three sed lines from the above).

If you like that idea, I'll file it as a separate issue.

---

I really like the `WORKDIR` change. Makes it super simple to insert a custom config.

---

Question: why set the bind address to `127.0.0.1`? I'm not entirely sure but depending on the docker networking mode this could cause the service to not be accessible from outside the container? With the sed statement is restricts all the services (sftp, webdav, httpd)...

The current version binds to all:

https://github.com/drakkan/sftpgo/blob/5b79379c9038359d2aecb7dda6f41db5533901e4/Dockerfile.alpine#L58

---

The instructions boil down to:
- mount `/srv/sftpgo` to handle persistent data
    - for me it would be:
        - mount `/srv/sftpgo/data` to handle user uploaded files
        - mount `/srv/sftpgo/backups` to handle backups
- mount `/var/lib/sftpgo` to persist host/keys and credentials - best done with a Docker Volume
- if you need to use a custom config, mount it to `/var/lib/sftpgo`, otherwise you can use environment variables (as per full configuration docs)

**Comment by drakkan** [comment]
> Looks great . I'm a little concerned that the default config file is so different from the standard sftpgo config. Perhaps (if you think it worthwhile) it might be good to change the default directories in sftpgo to be relative to some app data path by default.
> 
> E.g. `credentials`, `host_keys` (and local database paths) are relative to new `app_data_dir` config setting, which defaults to be equal to `config_dir` (for backwards compatibility). Then, we'd need to only change the new `app_data_dir` setting (which removes three sed lines from the above).
> 

thank you for noticing this. The paths are already relative we can you reduce the sed lines to these ones:

```
RUN sed -i "s|\"users_base_dir\": \"\",|\"users_base_dir\": \"/srv/sftpgo/data\",|" /etc/sftpgo/sftpgo.json && \
    sed -i "s|\"backups\"|\"/srv/sftpgo/backups\"|" /etc/sftpgo/sftpgo.json && \
    sed -i "s|\"bind_address\": \"127.0.0.1\",|\"bind_address\": \"\",|" /etc/sftpgo/sftpgo.json
```

> If you like that idea, I'll file it as a separate issue.
> 
> I really like the `WORKDIR` change. Makes it super simple to insert a custom config.
> 
> Question: why set the bind address to `127.0.0.1`? I'm not entirely sure but depending on the docker networking mode this could cause the service to not be accessible from outside the container? With the sed statement is restricts all the services (sftp, webdav, httpd)...
> 
> The current version binds to all:
> 
> https://github.com/drakkan/sftpgo/blob/5b79379c9038359d2aecb7dda6f41db5533901e4/Dockerfile.alpine#L58
> 
this address is only for the web admin interface, there is already a sed line to replace 127.0.0.1 with an empty string (all addresses)

> The instructions boil down to:
> 
> * mount `/srv/sftpgo` to handle persistent data
>   
>   * for me it would be:
>     
>     * mount `/srv/sftpgo/data` to handle user uploaded files
>     * mount `/srv/sftpgo/backups` to handle backups
> * mount `/var/lib/sftpgo` to persist host/keys and credentials - best done with a Docker Volume
> * if you need to use a custom config, mount it to `/var/lib/sftpgo`, otherwise you can use environment variables (as per full configuration docs)

thank you!

**Comment by drakkan** [comment]
Hi,

I pushed these changes, official images are now available on [Docker Hub](https://hub.docker.com/r/drakkan/sftpgo),

please open new issues or send a pull request if you think the images could be further improved, thank you
